PINIO BOX: Support configurable "ON duration" for box activation

elizarov opened the pull request:
By default, PINIO BOX facility is capable of turning on the configured PINIO when the associated box/mode is active and turning it off when it is no longer active. This PR adds an ability to optionally turn the pin on for a specified duration when the associated box/mode gets activated. The intended use-case for this feature is to control devices that use push-buttons which must be pressed for a certain duration to activate their functions.
An example that is presented at the end shows how this new feature can be used to control SMO 4K camera's power and recording in a more robust way than is possible via a regular PINIO BOX alone.

This PR introduces a new configuration option: "pinio_box_on_duration". It is an array of four numbers, similarly to "pinio_config" and "pinio_box" options. The default value of "pinio_box_on_duration" is "0,0,0,0". Zero value means that the corresponding PINIO works as usual -- it tracks the status of the corresponding box/mode as specified in "pinio_box". A non-zero value of "pinio_box_on_duration" specifies a duration in units of 100ms (e.g. 10 means 1 second). When the duration is configured, then the PINIO is turned on for the specified duration every time when the associated mode/box is activated.

### Example

For example, lets take an iFlight Beast F745 AIO FC (IFLIGHT_F745_AIO) and SMO 4K camera. Let's solder camera's power control wire to T7 pad on FC and camera's record control wire to R7 pad on FC, and reconfigure the corresponding pads for PINIO:

```
resource SERIAL_TX 7 NONE # was set to E08
resource SERIAL_RX 7 NONE # was set to E07
resource PINIO 1 E07      # R7 pad on FC
resource PINIO 2 E08      # T7 pad on FC
```

Now we configure PINIO_BOX to associate USER1 and USER2 boxes with PINIO 1 and PINIO 2.

```
set pinio_box = 40,41,255,255
```

Our goal is to turn on the power on SMO 4K and turn the recording on/off, which can be robustly done with 0.5 second press (pressing these buttons for too long activates their alternate functions and pressing them for too short does not get reliably recognized by the camera). So, we configure "pinio_box_on_duration" for both of those pins:

```
set pinio_box_on_duration = 5,5,0,0
```

Now we can go to "Modes" tab in configurator and configure USER1 (Camera recording) and USER2 (Camera power) modes. For example, we can take any AUX channel and assign a full range of this channel to USER2 box. This way, USER2 becomes active as soon as FC boots and, because of "pinio_box_on_duration" configuration, the corresponding pad is turned on for 0.5 seconds, powering on the attached camera. So, now the camera is powered on every time we attach a battery to FC (and will naturally be powered off when we detach the battery).

For recording control, for example, assume that we are using a high range of a three-position switch on AUX2 for PREARM. Let's set the same range of this AUX2 switch to USER1. This way, every time we move that switch to prearm, it will activate USER1 box, pushing a recording button for 0.5 seconds. You can flick the switch to prearm again to turn the recording off.

Note, that in order to work reliably, the switch must stay in the configured range for at least 20 ms to be reliably detected by PINIO BOX and trigger the USERx modes.

### Testing

The test binaries based on Betaflight 4.3.0-RC5 can be found here: https://github.com/elizarov/betaflight/releases/tag/pinio_box_on_duration-4.3.0-RC5

etracer65 commented:
There is no reason to create a new parameter group. Simply add the `pulse` array to the `pinioBoxConfig_t` structure.

Also "pulse" is probably a bad term since this is only a one-shot and not repeating as the name "pulse" implies. Maybe rename to "on duration" or something like that.

Lastly, I'm not sure 100ms is the appropriate value for the time units. Do we really need a maximum interval of 25.5 seconds? Maybe the scale should be 10ms to allow more flexibility for things that might need a really short duration.

elizarov commented:
@etracer65

> There is no reason to create a new parameter group. Simply add the `pulse` array to the `pinioBoxConfig_t` structure.

Thanks, will do. I've missed that it can be done this way. 

> Also "pulse" is probably a bad term since this is only a one-shot and not repeating as the name "pulse" implies. Maybe rename to "on duration" or something like that.

Makes sense. I'll update the parameter name to `pinio_box_on_duration`.

> Lastly, I'm not sure 100ms is the appropriate value for the time units. Do we really need a maximum interval of 25.5 seconds? Maybe the scale should be 10ms to allow more flexibility for things that might need a really short duration.

The problem is that `PINIOBOX` task runs at 20Hz, so it cannot generate delays smaller than 50ms. Would it be Ok to also update the task to run at 100Hz? As an additional bonus, it'll be able to "catch" more transient box activations, which could be noticeable when you manually operate a switch on a remote.

etracer65 commented:
> The problem is that `PINIOBOX` task runs at 20Hz, so it cannot generate delays smaller than 50ms. Would it be Ok to also update the task to run at 100Hz? As an additional bonus, it'll be able to "catch" more transient box activations, which could be noticeable when you manually operate a switch on a remote.

Good point. No, I don't think it would be appropriate to increase the `PINIOBOX` task frequency. The very limited need doesn't balance out the extra processing load.

In general I think this concept needs a rethink. The general idea of having a "momentary" pinio output is sound. But your use case of having it turn on momentarily on both the ON and OFF transitions of the `PINIOBOX` mode is problematic. Your workaround to have the 3-way switch mechanically pass through the mode on the way to arming or disarming is definitely a bit of a kludge. And as you noted it suffers from timing issues if the switch moves too quickly through the middle position. Since you could accomplish the same thing with automation directly in your radio (without the timing concerns) I'm not sure that putting this incomplete solution in Betaflight is worth it.

So while the initial concept of this PR to add a momentary output is valid and useful (for other use-cases), it's not really the full solution to your problem. So maybe a rethink to extend the pinio output behavior? Maybe add different operating modes like `ON` (the current behavior), momentary on start, momentary on start/finish, etc. Don't know and not making specific directions here. Maybe @jflyper already had some plans for this or some helpful input?

elizarov commented:
Let me share a bit of a back story so that we can fully discuss use-cases for this concept. My primary use case was turning on SMO 4K camera on FC boot. See, the camera takes a while to initialize, so it is extremely convenient to automatically press its power button as soon as FC is powered up (and this time I might not have even turned on the radio). Turning the recording on/off was not really a big problem. It already works decently via PINIO BOX on a switch and this "on for duration" feature adds a marginal improvement.

I've started by trying to find an existing BOX that is active for some time during FC boot but could not find one. In my initial prototype, I've just added a new BOX that does just this, but it looked hacky and very single-purposed. 

The next thing I've looked at was to actually extend the PINIO facility. The `pinio_config` has a lot of spare bits (only two are currently used), so it is natural to use additional bits to enable "turn on duration" feature. However, I was not able to find the right way to generate the delay inside pinoi code itself.

So, PINIO BOX was the next logical place, since it runs in a task. However, the 20Hz refresh rate of the PINIOBOX task is its weak point. Even on stock Betaflight, it is hard to reliably operate push buttons in the middle range of a 3-position switche, because of this slow rate. PINIOBOX task can simply miss an event if it happens relatively fast. 

I can offer to refactor pinio_box code and completely get rid of the PINIOBOX  task. He is what I see:
* The code for all box state transitions is well encapsulated, so state-change-tracking for PINIO BOX can be directly integrated there.
* The things that PINIOBOX task has to do (turning GPIO on and off) don't take a lot of CPU cycles. There's really not a lot of code in pinio_box, so if that code is called directly from the code that had initiated the change of the corresponding box state, then it will not result in a noticeable slowdown.

That would be, IMO, a useful improvement in itself (getting rid of an arbitrary 20Hz limit and having one less task). However, if the task is gone, then again, I don't know what would be the right way to generate a delay for the "on duration" feature. Alternatively, I can combine both approaches. I can move the change tracking code inside `rc_modes`, but use the task if "on duration" feature is used.

As for tracking OFF-ON box state transition in addition to OFF-ON transition. I was toying with the idea of doing exactly that to turn on the camera on each pre-arm and turn it off when pre-arm is deactivated. However, I've found that sometimes, for example, if you manually play with the camera, you might need to resync the recording state of the camera and that's where it is convenient to have an ability to generate one press of the camera's button with a switch on your radio. Getting it tightly bound to the pre-arm both ways can put you in a situation where your switch is already in pre-arm position and the camera is off, but you cannot do anything about it. 

However, if you see additional use-cases for tracking OFF-ON transitions instead or in addition to ON-OFF transitions, then the code can be naturally extended to cover those, too. In fact, if the implementation is done in such a way, that a box action for period that's shorter than the configured "on duration" generates a single IO signal, then you could still go ahead and couple your camerera recoding to PREARM box on both ON-OFF and OFF-ON transition, so in normal circumstances, your camera recording will start when you prearm and will stop when you turn off prearm, but you'll be able to always just quickly flick prearm switch to push the recording button on the camera. 

What do you think?

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

elizarov commented:
I've updated this PR, fixing all the low-handing issues identified so far:
* Renamed the setting to `pinio_box_on_duration`
* Updated implementation, moving all the configs into the `pinioBoxConfig_t` structure. This made the whole PR smaller, affecting less files.

SteveCEvans commented on src/main/io/piniobox.c:
```diff
@@ -38,39 +38,72 @@
+                    break;
+                case ON_DONE:
+                    if (desiredOn) {
+                        desiredOn = false; // on duration is over!
```
Do you intend this being a one-shot, which would make more sense, or having it re-triggerable as it is now. At present you'll get the desired one pulse followed by 50ms off, and repeat.

ctzsnooze commented:
@elizarov Please review the above comments, squash to latest master, confirm it works, and outline a testing method.

ctzsnooze commented:
@elizarov I'm curious if you're still supporting this PR?  If not, we should close it for now.

haslinghuis commented:
Closing as stalled
